src/compiler/checker.ts

```diff
@@ -29027,13 +29027,17 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
 
     function getTypeOfPropertyOfContextualType(type: Type, name: __String, nameType?: Type) {
         return mapType(type, t => {
-            if (isGenericMappedType(t) && !t.declaration.nameType) {
-                const constraint = getConstraintTypeFromMappedType(t);
-                const constraintOfConstraint = getBaseConstraintOfType(constraint) || constraint;
-                const propertyNameType = nameType || getStringLiteralType(unescapeLeadingUnderscores(name));
-                if (isTypeAssignableTo(propertyNameType, constraintOfConstraint)) {
-                    return substituteIndexedMappedType(t, propertyNameType);
-                }
+            if (everyContainedType(t, t => isGenericMappedType(t) && !t.declaration.nameType)) {

```

Comment:
Actually, thinking about it, shouldn't we just do this for each `isGenericMappedType(t) && !t.declaration.nameType` within an intersection, and not only do this substitution if every member is such? (And still do the other `StructuredType` behaviors for other intersection members). This way a `Results<T> & Errors<E> & {x?: string}` still works.

Reply:
Right, so I actually... have another PR that does exactly that 🤣 This one here is intentionally conservative when it comes to the scope of the change, especially since some other changes (outside of `getTypeOfPropertyOfContextualType`) had to be made to support this use case. So it's not like that other PR completely replaces this one.

The PR in question is this: https://github.com/microsoft/TypeScript/pull/52095 . Actually, one version of it was already merged into 4.8-beta but it was reverted cause `@reduxjs/toolkit` regressed with it. However, it only regressed because of how I've handled intersections of types with concrete properties with types with index signatures. 

In the original version of this PR (https://github.com/microsoft/TypeScript/pull/48668) I decided to prefer concrete properties over index signatures, so (IIRC) if one of the intersected members had a concrete property for the requested one then index signatures on other intersected members were skipped over. This **wasn't strictly needed** to fix the issue that I was fixing but it was something that made sense to me at the time.

Either way... if you could take a look at that other PR (https://github.com/microsoft/TypeScript/pull/52095), I would highly appreciate it. There is some funky stuff going on there with index signatures and that part is not correct - the problem is that I'm not sure how to adjust it since I'm not sure what exact effect are we aiming for there. Mixing index signatures with concrete properties through intersections is weird (like `{ foo: number } & { [key: string]: boolean }`). This particular place in the code doesn't exactly deal with concrete types either... so perhaps a guiding principle should be that properties should be pulled from types that can still resolve to concrete properties? I'm not sure how checking that would look like though - so perhaps we should just ignore it and always pull from index signatures?

Reply:
To keep the spirit of minimal changes I would probably prefer to keep them separate (if you prefer to merge them, then I won't oppose it though, at the end of the day I'd like both improvements to find their way into the language :P). I don't mind this one staying on the sidelines and waiting for the other one to land though. I would love for the other one to get some traction in terms of being reviewed 😜

Reply:
I think that the common improvements need to be on both PRs (and the conflicts can be resolved later), or both PRs should be merged, yeah.